client/protocols/hinting.py

```diff
@@ -0,0 +1,185 @@
+
+        After more attempts, ask if students would like hints. If so, query
+        the server.
+        """
+        if self.args.local:
+            return
+
+        if 'analytics' not in messages:
+            log.info('Analytics Protocol is required for hint generation')
+            return
+        if 'file_contents' not in messages:
+            log.info('File Contents needed to generate hints')
+            return
+
+        messages['hinting'] = {}
+        history = messages['analytics']['history']
+        questions = history['questions']
+        current_q = history['question']
+
+        for question in current_q:
+            if question not in questions:
+                continue
+            stats = questions[question]
+            messages['hinting'][question] = {'prompts': {}, 'reflection': {}}
+            hint_info = messages['hinting'][question]
+
+            # Only prompt
+            if (stats['solved'] and stats['attempts'] > 6):
+                hint_info['elgible'] = False
+                if self.args.question:
+                    # Only prompt for reflection with question specified.
+                    log.info('Giving reflection response on %s', question)
+                    reflection = random.choice(SOLVE_SUCCESS_MSG)
+                    if not confirm("Nice work! Could you answer a quick question"
+                                   " about how you approached this question?"):
+                        hint_info['reflection']['accept'] = False
+                    else:
+                        hint_info['reflection']['accept'] = True
+                    prompt_user(reflection, hint_info)
+            elif (stats['attempts'] < 5):
+                log.info("Question %s is not elgible: Attempts: %s, Solved: %s",
+                         question, stats['attempts'], stats['solved'])
+                hint_info['elgible'] = False
+            else:
+                hint_info['elgible'] = not stats['solved']
+
+            if not hint_info['elgible']:
+                continue
+            log.info('Prompting for hint on %s', question)
+
+            if confirm("Check for hints on {}?".format(question)):
+                hint_info['accept'] = True
+                print("Thinking... (could take upto 15 seconds)", end='')
+                try:
+                    response = self.query_server(messages, question)
+                    hint_info['response'] = response
+
+                    hint = response['message']
+                    pre_prompt = response['pre-prompt']
+                    post_prompt = response['post-prompt']
+                    log.info("Hint server response: {}".format(response))
+                    if not hint and not pre_prompt:
+                        print("\nSorry. No hints found for {}".format(question))
+                        continue
+
+                    if pre_prompt:
+                        print("\r-- While we wait, respond to this question."
```

Comment:
What exactly are we waiting for? Won't we only get to this code path when `query_Server` finishes executing?